### PR TITLE
Charlie docker aspnetcore post

### PR DESCRIPTION
**Pull request by Quorralyne**
This blog post has the following features:

- [ ] A GitHub Repository with a polished README
- [X] A title that's approved by Dev Marketing
- [X] A URL that's approved by Dev Marketing
- [X] Rendered locally and confirmed that no Markdown typos exist

https://docs.google.com/document/d/11lbGZhfzqI0U1XLRyEdaeMTvtbKfag8X3Xo3tuRAuGM/edit?pli=1#

### CODE REVIEW

File: _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md [hunk 1]
```diff
@@ -0,0 +1,343 @@
+---
+layout: blog_post
+title: "Build a Simple .NET Core App on Docker"
+author: charlieholland
+description: "Are you in tech? Getting ready to travel? These travel tips will save you time and reduce stress on your next conference or work trip."
```
Comment (mraible) [inline _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md]: Description needs to be updated to match the post.

File: _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+tags: [ dotnetcore, docker, dotnet, .netcore, csharp ]
+tweets:
+- "Ever wonder what you can with #Docker + #AspNetCore? Check it out! →"
+- "Hey .NET devs, we've got a simple #Docker tutorial for your #AspNetCore apps! →"
+- "Traveling for work or to a conference? Make your journey a breeze with these tips →"
+image: blog/featured/okta-dotnet-mouse-down.jpg
+---
+
+Wouldn’t it be great if stuff just worked? Especially in the ever-changing world of software. Chasing dependency issues and debugging arcane operating system errors - not a good use of time.
```
Comment (mraible) [inline _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md]: Fix single quote so it's not a curly quote.

File: _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+
+One important aspect of "stuff just works" -- reliability. Recently, the software community has made strides in test-driven development and continuous integration processes to drive up quality, and of course, that improves reliability. But it can only go so far. Operating systems perform many functions and incorporate many features, so running them reliably is a huge task. The core secret to improving reliability is fewer moving parts. In this post, I’ll cover how you can use Docker to run your application in an isolated, minimal environment with fewer moving parts.
+
+## Sample App Dependencies: ASP.Net Core and Docker Packages
+
+To build an app in Docker, first we need an app to Dockerize. This tutorial uses the  ASP.Net Core project from a previous blogpost on [Adding Login to Your ASP.NET Core MVC App](https://developer.okta.com/blog/2018/10/29/add-login-to-you-aspnetcore-app).
```
Comment (mraible) [inline _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md]: Links to blog posts should be relative.

File: _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+
+## Update the Settings in the ASP.NET Core App
+
+The sample application you created has identity management configuration stored in `appsettings.json`, unless you cloned the repo above rather than working through the complete post. In real life, you shouldn't store this configuration in your source code for security reasons. In this post, we will demonstrate how to reliably pass dynamic configuration to your application, and close this security gap. Start by removing the settings.
+
+Edit `appsettings.json` to remove:
+
+```json
+"Okta": {
+"ClientId": "{OktaClientId}",
```
Comment (mraible) [inline _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md]: I'd recommend indenting the JSON.

File: _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+"ClientSecret": "{OktaClientSecret}",
+"Domain": "https://dev-107646.okta.com"
```
Comment (mraible) [inline _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md]: dev-107646 > {yourOktaDomain}

File: _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+WORKDIR /src
+COPY ["OktaMvcLogin.csproj", "./"]
+RUN dotnet restore "./OktaMvcLogin.csproj"
+COPY . .
+RUN dotnet build "OktaMvcLogin.csproj" -c Release -o /app
+```
+
+The uppercase words are Docker commands. There aren't many of them, and you can find the details of them all at [Docker's website](http://docker.io).
+
+`FROM` tells Docker which image you want to use for your container. An image is a compressed file system snapshot. Also, the result of building a `Dockerfile` is a new image. So, one way to look at a `Dockerfile` is as a series of transformations that convert one image into another image that includes your application.
```
Comment (mraible) [inline _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md]: Is this supposed to be a bullet list?

File: _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+
+So, in plain English - this `Dockerfile` is based on the `dotnet/core/sdk` image hosted at `mcr.microsoft.com`. Docker copies the `.csproj` file from your local working directory to create your image and `dotnet restore` restores all the referenced packages. Once that's done, Docker copies the remaining files from your working directory, then `dotnet build` creates a **Release** build at `/app`.
+
+## Manage Dependencies Efficiently with Docker
+
+Reading this, you may be thinking, why bother to copy the project file and run `restore` before copying the source code and running `build`. Why not copy everything then build and restore in one step? The answer is caching. Every time a `Dockerfile` modifies the docker image, Docker creates a snapshot.  If you copy a file or run a command to install a package, Docker captures the differences in a new snapshot. Docker then caches and reuses the snapshots if the image hasn't changed. So, by restoring dependencies as a separate step, the image snapshot can be reused for every build, as long as the dependencies haven't changed. This process speeds up the build considerably since downloading dependencies can take some time.
```
Comment (mraible) [inline _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md]: Add ? after "the source code and running `build`"

File: _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+
+## Run the ASP.NET Core App in a Docker Container
+
+As mentioned above, a `Dockerfile` can be considered a series of filesystem transformations. Your current file transforms the Microsoft-provided SDK container into a new container with both the Microsoft SDK and a release build of your application stored at `/app`.
+
+Check this out::
```
Comment (mraible) [inline _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md]: Check > Try

File: _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+
+```bash
+# Build an image using the Dockerfile in the current directory
+docker build --target build -t oktamvclogin .
+# Run the image, executing the command 'ls /app'
+docker run -it oktamvclogin ls /app
```
Comment (mraible) [inline _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md]: Remove blank line after this one.

File: _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+COPY --from=publish /app .
+ENTRYPOINT ["dotnet", "OktaMvcLogin.dll"]
+```
+
+You will see a few `FROM` commands, each with an `AS` clause. This syntax provides multi-stage builds, the key to getting rid of unnecessary dependencies. In plain English, your build process is now:
+
+1. Use the SDK image to create a release build of the application. Call this stage 'build'
+2. Use the 'build' stage image to publish the application to the 'app' folder. Call this stage 'publish'
+3. Download the Microsoft-provided ASP.NET core image, which has only runtime components. Call this stage 'base'
+4. Using the 'base' stage image, copy the contents of the 'app' directory from the 'publish' stage. Call this stage 'final'
```
Comment (mraible) [inline _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md]: I would add periods at the end of each item since the previous sentence has a period.

File: _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+This time the result will be a bit healthier:
+
+```bash
+Hosting environment: Production
+Content root path: /app
+Now listening on: 'http://[::]:80'
+Application started. Press Ctrl+C to shut down.
+```
+
+>**Note**: you may also see a 'No XML Encryptor' warning. You can ignore that for this walkthrough.
```
Comment (mraible) [inline _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md]: I would remove the > to make this a blockquote and change the note to be all caps. **NOTE:**

File: _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+docker run -e ASPNETCORE_Okta__ClientId="{yourClientId}-" \
+-e ASPNETCORE_Okta__ClientSecret="{yourClientSecret}" \
+-e ASPNETCORE_Okta__Domain="{yourOktaDomain}" \
+-p 5001:80 \
+oktamvclogin
+```
+
+Now, if you open a browser and go to `http://localhost:5001` (because you mapped port 5001 to port 80 in your container), Et voila!
+
+>**Note**: this approach is suitable for development. However, for production workloads, Docker offers a comprehensive set of options designed for managing virtual networks. For more information see [the networking overview in Docker’s documentation](https://docs.docker.com/network/).
```
Comment (mraible) [inline _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md]: Same comment as above blockquote.

File: _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+
+## Configure SSL/TLS for Your Docker Image
+
+If you click on the 'Login' link in your application, chances are you'll get an error from Okta with a message:
```
Comment (mraible) [inline _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md]: 'Login' > **Login**

File: _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+-e ASPNETCORE_Kestrel__Certificates__Default__Password="{yourExportPassword}" \
+-p 5001:443 \
+oktamvclogin
+```
+
+Notice the additional environment variable with the certificate export password and also, that the port mapping has changed from port 80 to port 443.
+
+You can now navigate to `https://localhost:5001`, and this time, you'll be able to log in and use the sample application correctly.
+
+>Note: Since you're using a self-signed certificate, your browser may display a warning page. You can safely ignore this warning.
```
Comment (mraible) [inline _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md]: Make note consistent with the ones above.

File: _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+
+## Not the Best Way to Start Your Docker Container
+
+Converting an application to run in Docker is relatively straightforward and offers significant benefits. However, passing a whole load of configuration to `docker run` isn't particularly user-friendly and is error-prone. Thankfully, the good folks at Docker have already come up with a solution to this problem - `docker-compose`.
+
+Using Docker Compose is pretty straightforward. Create a file named `docker-compose.yml` in the same folder as your `Dockerfile` and source code and add the following:
+
+```yaml
```
Comment (mraible) [inline _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md]: This YAML is incorrect. It needs to be formatted properly.

File: _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md [hunk 16]
```diff
@@ -0,0 +1,343 @@
...
+Now, by including a `Dockerfile` along with your source code, any developer can build your app -- reliably. They can build it in any environment, as long as they have Docker installed, and it will work in precisely the same way. Every time, without exception. No dependency issues, no operating system hassles. It will just work. Not only that, but you can deploy the same container directly to production without any further modification.
+
+## Learn More About Docker and ASP.NET Core
+
+If you want to learn more about ASP.NET Core, Okta, or Docker, check out these other posts!
+
+* [A Developer's Guide to Docker - A Gentle Introduction](https://developer.okta.com/blog/2017/05/10/developers-guide-to-docker-part-1)
+* [A Developer's Guide to Docker - The Dockerfile](https://developer.okta.com/blog/2017/08/28/developers-guide-to-docker-part-2)
+* [A Developer's Guide to Docker - Docker Compose](https://developer.okta.com/blog/2017/10/11/developers-guide-to-docker-part-3)
+* [Add Login to Your ASP.NET Core MVC App](https://developer.okta.com/blog/2018/10/29/add-login-to-you-aspnetcore-app)
```
Comment (mraible) [inline _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md]: Links to blog posts are usually relative.

File: _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md (hunk 15 above)
Comment (Quorralyne) [reply _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md]: How should it be formatted?

File: _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md [hunk 17]
```diff
@@ -0,0 +1,342 @@
...
+image: blog/featured/okta-dotnet-mouse-down.jpg
+---
+
+Wouldn't it be great if stuff just worked? Especially in the ever-changing world of software. Chasing dependency issues and debugging arcane operating system errors - not a good use of time.
+
+One important aspect of "stuff just works" -- reliability. Recently, the software community has made strides in test-driven development and continuous integration processes to drive up quality, and of course, that improves reliability. But it can only go so far. Operating systems perform many functions and incorporate many features, so running them reliably is a huge task. The core secret to improving reliability is fewer moving parts. In this post, I'll cover how you can use Docker to run your application in an isolated, minimal environment with fewer moving parts.
+
+## Sample App Dependencies: ASP.Net Core and Docker Packages
+
+To build an app in Docker, first we need an app to Dockerize. This tutorial uses the  ASP.Net Core project from a previous blogpost on [Adding Login to Your ASP.NET Core MVC App](blog/2018/10/29/add-login-to-you-aspnetcore-app).
```
Comment (mraible) [inline _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md]:
```suggestion
To build an app in Docker, first we need an app to Dockerize. This tutorial uses the  ASP.Net Core project from a previous blog post on [Adding Login to Your ASP.NET Core MVC App](/blog/2018/10/29/add-login-to-you-aspnetcore-app).
```

File: _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md [hunk 18]
```diff
@@ -0,0 +1,342 @@
...
+That blogpost shows you how to build a simple .Net Core application that uses Okta for Identity Management. You should work through that blog post, or at very least read it and clone the repo.
```
Comment (mraible) [inline _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md]:
```suggestion
That blog post shows you how to build a simple .Net Core application that uses Okta for Identity Management. You should work through that blog post, or at very least read it and clone the repo.
```

File: _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md [hunk 19]
```diff
@@ -0,0 +1,342 @@
...
+
+You'll also need:
```
Comment (mraible) [inline _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md]: I would change this to match the Prerequisites block like https://developer.okta.com/blog/2019/09/11/angular-mongodb has.

File: _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md (hunk 15 above)
Comment (mraible) [reply _source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md]: Here's a guess based on other Docker compile files I've seen:

```yaml
version: "3"
services:
  web:
    build: .
    image: oktamvclogin
    ports:
      - "5001:443"
    environment:
      - ASPNETCORE_Okta__ClientId: "{yourClientId}-"
      - ASPNETCORE_Okta__ClientSecret: "{yourClientSecret}"
      - ASPNETCORE_Okta__Domain: "{yourOktaDomain}"
      - ASPNETCORE_Kestrel__Certificates__Default__Password: "{yourExportPassword}"
```
